File: storage/src/visibility.rs
//! Tuple visibility rules for MVCC.
//!
//! This module implements PostgreSQL's tuple visibility logic as defined in
//! `src/backend/access/heap/heapam_visibility.c`. The visibility rules determine
//! whether a tuple is visible to a given transaction based on:
//!
//! - The tuple's `xmin` (inserting transaction)
//! - The tuple's `xmax` (deleting transaction, if any)
//! - The current transaction's snapshot
//! - Hint bits in the tuple header
//!
//! # MVCC Basics
//!
//! PostgreSQL uses Multi-Version Concurrency Control (MVCC) to provide
//! transaction isolation without read locks. Each tuple has:
//!
//! - `xmin`: The transaction ID that inserted this tuple version
//! - `xmax`: The transaction ID that deleted/updated this tuple (0 if not deleted)
//!
//! A tuple is visible if:
//! 1. `xmin` is committed and visible to the snapshot
//! 2. `xmax` is either invalid, aborted, or not visible to the snapshot
//!
//! # Hint Bits
//!
//! To avoid repeated transaction status lookups, PostgreSQL sets "hint bits"
//! on tuples after determining their visibility status:
//!
//! - `HEAP_XMIN_COMMITTED`: xmin is known to be committed
//! - `HEAP_XMIN_INVALID`: xmin is known to be aborted
//! - `HEAP_XMAX_COMMITTED`: xmax is known to be committed
//! - `HEAP_XMAX_INVALID`: xmax is known to be invalid/aborted

use crate::tuple::{HeapTupleHeader, InfomaskFlags};
use crate::types::{
    CommandId, TransactionId, FIRST_NORMAL_TRANSACTION_ID, FROZEN_TRANSACTION_ID,
    INVALID_TRANSACTION_ID,
};

/// Transaction status.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum TransactionStatus {
    /// Transaction is still in progress.
    InProgress,
    /// Transaction has committed.
    Committed,
    /// Transaction has aborted.
    Aborted,
}

/// A snapshot represents a consistent view of the database at a point in time.
///
/// Matches PostgreSQL's `SnapshotData` from `snapshot.h`.
#[derive(Debug, Clone)]
pub struct Snapshot {
    /// The transaction ID of the current transaction.
    pub xid: TransactionId,
    /// All transaction IDs >= this are invisible.
    pub xmax: TransactionId,
    /// All transaction IDs < this are visible (if committed).
    pub xmin: TransactionId,
    /// Transaction IDs of in-progress transactions (between xmin and xmax).
    pub xip: Vec<TransactionId>,
    /// Current command ID within the transaction.
    pub curcid: CommandId,
}

impl Snapshot {
    /// Create a new snapshot.
    #[must_use]
    pub fn new(xid: TransactionId, xmin: TransactionId, xmax: TransactionId) -> Self {
        Self {
            xid,
            xmax,
            xmin,
            xip: Vec::new(),
            curcid: 0,
        }
    }

    /// Create a snapshot that sees all committed transactions.
    ///
    /// This is useful for system operations that need to see everything.
    #[must_use]
    pub fn any() -> Self {
        Self {
            xid: FIRST_NORMAL_TRANSACTION_ID,
            xmax: TransactionId::MAX,
            xmin: FIRST_NORMAL_TRANSACTION_ID,
            xip: Vec::new(),
            curcid: 0,
        }
    }

    /// Check if a transaction ID is visible in this snapshot.
    ///
    /// A transaction is visible if:
    /// - It's less than xmin (definitely committed before snapshot)
    /// - It's between xmin and xmax and not in the xip list
    /// - It's the current transaction
    #[must_use]
    pub fn is_visible(&self, xid: TransactionId) -> bool {
        // Special transaction IDs
        if xid == INVALID_TRANSACTION_ID {
            return false;
        }
        if xid == FROZEN_TRANSACTION_ID {
            return true;
        }

        // Current transaction's own changes are visible
        if xid == self.xid {
            return true;
        }

        // Transaction IDs >= xmax are not visible
        if xid >= self.xmax {
            return false;
        }

        // Transaction IDs < xmin are visible (if committed)
        if xid < self.xmin {
            return true;
        }

        // Check if in the in-progress list
        !self.xip.contains(&xid)
    }

    /// Add an in-progress transaction to the snapshot.
    pub fn add_in_progress(&mut self, xid: TransactionId) {
        if !self.xip.contains(&xid) {
            self.xip.push(xid);
        }
    }
}

impl Default for Snapshot {
    fn default() -> Self {
        Self::any()
    }
}

/// Result of a visibility check.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum VisibilityResult {
    /// Tuple is visible to the snapshot.
    Visible,
    /// Tuple is not visible (inserted by uncommitted/aborted transaction).
    Invisible,
    /// Tuple is deleted and not visible.
    Deleted,
    /// Tuple is being deleted by current transaction.
    BeingDeleted,
    /// Tuple was inserted by current transaction.
    InsertedBySelf,
    /// Tuple was deleted by current transaction.
    DeletedBySelf,
}

impl VisibilityResult {
    /// Check if the tuple should be returned to the user.
    #[must_use]
    pub fn is_visible(self) -> bool {
        matches!(self, Self::Visible | Self::InsertedBySelf)
    }
}

/// Transaction status oracle - provides transaction commit status.
///
/// In a real implementation, this would consult pg_xact (clog).
/// For now, we use a simple in-memory map.
pub trait TransactionOracle {
    /// Get the status of a transaction.
    fn get_status(&self, xid: TransactionId) -> TransactionStatus;

    /// Check if a transaction is committed.
    fn is_committed(&self, xid: TransactionId) -> bool {
        self.get_status(xid) == TransactionStatus::Committed
    }

    /// Check if a transaction is aborted.
    fn is_aborted(&self, xid: TransactionId) -> bool {
        self.get_status(xid) == TransactionStatus::Aborted
    }

    /// Check if a transaction is in progress.
    fn is_in_progress(&self, xid: TransactionId) -> bool {
        self.get_status(xid) == TransactionStatus::InProgress
    }
}

/// Simple in-memory transaction oracle for testing.
#[derive(Debug, Default)]
pub struct SimpleTransactionOracle {
    /// Map of transaction ID to status.
    status: std::collections::HashMap<TransactionId, TransactionStatus>,
}

impl SimpleTransactionOracle {
    /// Create a new oracle.
    #[must_use]
    pub fn new() -> Self {
        Self::default()
    }

    /// Set the status of a transaction.
    pub fn set_status(&mut self, xid: TransactionId, status: TransactionStatus) {
        self.status.insert(xid, status);
    }

    /// Mark a transaction as committed.
    pub fn commit(&mut self, xid: TransactionId) {
        self.set_status(xid, TransactionStatus::Committed);
    }

    /// Mark a transaction as aborted.
    pub fn abort(&mut self, xid: TransactionId) {
        self.set_status(xid, TransactionStatus::Aborted);
    }
}

impl TransactionOracle for SimpleTransactionOracle {
    fn get_status(&self, xid: TransactionId) -> TransactionStatus {
        // Special cases
        if xid == INVALID_TRANSACTION_ID {
            return TransactionStatus::Aborted;
        }
        if xid == FROZEN_TRANSACTION_ID {
            return TransactionStatus::Committed;
        }

        self.status
            .get(&xid)
            .copied()
            .unwrap_or(TransactionStatus::InProgress)
    }
}

/// Check if a heap tuple is visible to a snapshot.
///
/// This is the main visibility check function, implementing the logic from
/// `HeapTupleSatisfiesMVCC` in PostgreSQL.
///
/// # Arguments
///
/// * `header` - The tuple header to check
/// * `snapshot` - The snapshot to check visibility against
/// * `oracle` - Transaction status oracle
///
/// # Returns
///
/// The visibility result for the tuple.
pub fn heap_tuple_satisfies_mvcc<O: TransactionOracle>(
    header: &HeapTupleHeader,
    snapshot: &Snapshot,
    oracle: &O,
) -> VisibilityResult {
    let infomask = header.infomask();

    // Check xmin (inserting transaction)
    if !check_xmin_visible(header, snapshot, oracle, infomask) {
        return VisibilityResult::Invisible;
    }

    // Check if inserted by current transaction
    if header.t_xmin == snapshot.xid {
        // Check command ID for visibility within same transaction
        if header.t_cid >= snapshot.curcid {
            return VisibilityResult::Invisible;
        }

        // Check xmax
        if header.t_xmax == INVALID_TRANSACTION_ID || infomask.contains(InfomaskFlags::HEAP_XMAX_INVALID) {
            return VisibilityResult::InsertedBySelf;
        }

        if header.t_xmax == snapshot.xid {
            return VisibilityResult::DeletedBySelf;
        }
    }

    // Check xmax (deleting transaction)
    check_xmax_visibility(header, snapshot, oracle, infomask)
}

/// Check if xmin makes the tuple visible.
fn check_xmin_visible<O: TransactionOracle>(
    header: &HeapTupleHeader,
    snapshot: &Snapshot,
    oracle: &O,
    infomask: InfomaskFlags,
) -> bool {
    let xmin = header.t_xmin;

    // Invalid xmin means tuple was never properly inserted
    if xmin == INVALID_TRANSACTION_ID {
        return false;
    }

    // Frozen tuples are always visible
    if xmin == FROZEN_TRANSACTION_ID {
        return true;
    }

    // Check hint bits first
    if infomask.contains(InfomaskFlags::HEAP_XMIN_COMMITTED) {
        // xmin is committed, check if visible to snapshot
        return snapshot.is_visible(xmin);
    }

    if infomask.contains(InfomaskFlags::HEAP_XMIN_INVALID) {
        // xmin is aborted
        return false;
    }

    // No hint bits, need to check transaction status
    match oracle.get_status(xmin) {
        TransactionStatus::Committed => snapshot.is_visible(xmin),
        TransactionStatus::Aborted => false,
        TransactionStatus::InProgress => {
            // In-progress transaction - only visible if it's our own
            xmin == snapshot.xid
        }
    }
}

/// Check xmax visibility and return final result.
fn check_xmax_visibility<O: TransactionOracle>(
    header: &HeapTupleHeader,
    snapshot: &Snapshot,
    oracle: &O,
    infomask: InfomaskFlags,
) -> VisibilityResult {
    let xmax = header.t_xmax;

    // No deleting transaction
    if xmax == INVALID_TRANSACTION_ID || infomask.contains(InfomaskFlags::HEAP_XMAX_INVALID) {
        return VisibilityResult::Visible;
    }

    // Check hint bits
    if infomask.contains(InfomaskFlags::HEAP_XMAX_COMMITTED) {
        // xmax is committed - tuple is deleted
        if snapshot.is_visible(xmax) {
            return VisibilityResult::Deleted;
        }
        // Deletion not visible to this snapshot
        return VisibilityResult::Visible;
    }

    // Check if being deleted by current transaction
    if xmax == snapshot.xid {
        return VisibilityResult::BeingDeleted;
    }

    // Check transaction status
    match oracle.get_status(xmax) {
        TransactionStatus::Committed => {
            if snapshot.is_visible(xmax) {
                VisibilityResult::Deleted
            } else {
                VisibilityResult::Visible
            }
        }
        TransactionStatus::Aborted => {
            // Deletion was aborted, tuple is visible
            VisibilityResult::Visible
        }
        TransactionStatus::InProgress => {
            // Deletion in progress by another transaction
            VisibilityResult::Visible
        }
    }
}

/// Set hint bits on a tuple header based on visibility check results.
///
/// This should be called after a visibility check to avoid repeated lookups.
pub fn set_hint_bits<O: TransactionOracle>(header: &mut HeapTupleHeader, oracle: &O) {
    let mut infomask = header.infomask();

    // Set xmin hints
    if !infomask.contains(InfomaskFlags::HEAP_XMIN_COMMITTED)
        && !infomask.contains(InfomaskFlags::HEAP_XMIN_INVALID)
    {
        match oracle.get_status(header.t_xmin) {
            TransactionStatus::Committed => {
                infomask.insert(InfomaskFlags::HEAP_XMIN_COMMITTED);
            }
            TransactionStatus::Aborted => {
                infomask.insert(InfomaskFlags::HEAP_XMIN_INVALID);
            }
            TransactionStatus::InProgress => {}
        }
    }

    // Set xmax hints
    if header.t_xmax != INVALID_TRANSACTION_ID
        && !infomask.contains(InfomaskFlags::HEAP_XMAX_COMMITTED)
        && !infomask.contains(InfomaskFlags::HEAP_XMAX_INVALID)
    {
        match oracle.get_status(header.t_xmax) {
            TransactionStatus::Committed => {
                infomask.insert(InfomaskFlags::HEAP_XMAX_COMMITTED);
            }
            TransactionStatus::Aborted => {
                infomask.insert(InfomaskFlags::HEAP_XMAX_INVALID);
            }
            TransactionStatus::InProgress => {}
        }
    }

    header.set_infomask(infomask);
}

#[cfg(test)]
mod tests {
    use super::*;

    fn make_tuple(xmin: TransactionId, xmax: TransactionId) -> HeapTupleHeader {
        let mut header = HeapTupleHeader::new(xmin, 0);
        header.t_xmax = xmax;
        // If xmax is set, clear the HEAP_XMAX_INVALID flag
        if xmax != INVALID_TRANSACTION_ID {
            let mut flags = header.infomask();
            flags.remove(InfomaskFlags::HEAP_XMAX_INVALID);
            header.set_infomask(flags);
        }
        header
    }

    #[test]
    fn snapshot_visibility_basic() {
        let snapshot = Snapshot::new(100, 50, 150);

        // Frozen is always visible
        assert!(snapshot.is_visible(FROZEN_TRANSACTION_ID));

        // Invalid is never visible
        assert!(!snapshot.is_visible(INVALID_TRANSACTION_ID));

        // Current transaction is visible
        assert!(snapshot.is_visible(100));

        // Old committed transactions are visible
        assert!(snapshot.is_visible(40));

        // Future transactions are not visible
        assert!(!snapshot.is_visible(200));
    }

    #[test]
    fn snapshot_with_in_progress() {
        let mut snapshot = Snapshot::new(100, 50, 150);
        snapshot.add_in_progress(75);
        snapshot.add_in_progress(80);

        // In-progress transactions are not visible
        assert!(!snapshot.is_visible(75));
        assert!(!snapshot.is_visible(80));

        // Other transactions in range are visible
        assert!(snapshot.is_visible(70));
        assert!(snapshot.is_visible(90));
    }

    #[test]
    fn simple_oracle() {
        let mut oracle = SimpleTransactionOracle::new();

        // Default is in-progress
        assert!(oracle.is_in_progress(100));

        oracle.commit(100);
        assert!(oracle.is_committed(100));
        assert!(!oracle.is_in_progress(100));

        oracle.abort(200);
        assert!(oracle.is_aborted(200));
    }

    #[test]
    fn visible_committed_tuple() {
        let mut oracle = SimpleTransactionOracle::new();
        oracle.commit(50);

        let header = make_tuple(50, INVALID_TRANSACTION_ID);
        let snapshot = Snapshot::new(100, 40, 150);

        let result = heap_tuple_satisfies_mvcc(&header, &snapshot, &oracle);
        assert_eq!(result, VisibilityResult::Visible);
    }

    #[test]
    fn invisible_uncommitted_tuple() {
        let oracle = SimpleTransactionOracle::new();

        let header = make_tuple(50, INVALID_TRANSACTION_ID);
        let snapshot = Snapshot::new(100, 40, 150);

        let result = heap_tuple_satisfies_mvcc(&header, &snapshot, &oracle);
        assert_eq!(result, VisibilityResult::Invisible);
    }

    #[test]
    fn invisible_aborted_tuple() {
        let mut oracle = SimpleTransactionOracle::new();
        oracle.abort(50);

        let header = make_tuple(50, INVALID_TRANSACTION_ID);
        let snapshot = Snapshot::new(100, 40, 150);

        let result = heap_tuple_satisfies_mvcc(&header, &snapshot, &oracle);
        assert_eq!(result, VisibilityResult::Invisible);
    }

    #[test]
    fn deleted_tuple() {
        let mut oracle = SimpleTransactionOracle::new();
        oracle.commit(50);
        oracle.commit(60);

        let header = make_tuple(50, 60);
        let snapshot = Snapshot::new(100, 40, 150);

        let result = heap_tuple_satisfies_mvcc(&header, &snapshot, &oracle);
        assert_eq!(result, VisibilityResult::Deleted);
    }

    #[test]
    fn tuple_with_uncommitted_delete() {
        let mut oracle = SimpleTransactionOracle::new();
        oracle.commit(50);
        // xmax (60) is still in progress

        let header = make_tuple(50, 60);
        let snapshot = Snapshot::new(100, 40, 150);

        let result = heap_tuple_satisfies_mvcc(&header, &snapshot, &oracle);
        assert_eq!(result, VisibilityResult::Visible);
    }

    #[test]
    fn tuple_with_aborted_delete() {
        let mut oracle = SimpleTransactionOracle::new();
        oracle.commit(50);
        oracle.abort(60);

        let header = make_tuple(50, 60);
        let snapshot = Snapshot::new(100, 40, 150);

        let result = heap_tuple_satisfies_mvcc(&header, &snapshot, &oracle);
        assert_eq!(result, VisibilityResult::Visible);
    }

    #[test]
    fn own_insert_visible() {
        let oracle = SimpleTransactionOracle::new();
        // Transaction 100 is current, so in-progress

        let mut header = make_tuple(100, INVALID_TRANSACTION_ID);
        header.t_cid = 0;

        let mut snapshot = Snapshot::new(100, 40, 150);
        snapshot.curcid = 5;

        let result = heap_tuple_satisfies_mvcc(&header, &snapshot, &oracle);
        assert_eq!(result, VisibilityResult::InsertedBySelf);
    }

    #[test]
    fn own_insert_not_visible_same_command() {
        let oracle = SimpleTransactionOracle::new();

        let mut header = make_tuple(100, INVALID_TRANSACTION_ID);
        header.t_cid = 5;

        let mut snapshot = Snapshot::new(100, 40, 150);
        snapshot.curcid = 5; // Same command ID

        let result = heap_tuple_satisfies_mvcc(&header, &snapshot, &oracle);
        assert_eq!(result, VisibilityResult::Invisible);
    }

    #[test]
    fn own_delete() {
        let mut oracle = SimpleTransactionOracle::new();
        oracle.commit(50);

        let mut header = make_tuple(50, 100);
        header.t_cid = 0;

        let snapshot = Snapshot::new(100, 40, 150);

        let result = heap_tuple_satisfies_mvcc(&header, &snapshot, &oracle);
        assert_eq!(result, VisibilityResult::BeingDeleted);
    }

    #[test]
    fn hint_bits_set() {
        let mut oracle = SimpleTransactionOracle::new();
        oracle.commit(50);
        oracle.abort(60);

        let mut header = make_tuple(50, 60);

        // Initially no hint bits
        assert!(!header.xmin_committed());
        assert!(!header.xmax_invalid());

        set_hint_bits(&mut header, &oracle);

        // Now hint bits should be set
        assert!(header.xmin_committed());
        assert!(header.xmax_invalid());
    }

    #[test]
    fn frozen_tuple_always_visible() {
        let oracle = SimpleTransactionOracle::new();

        let header = make_tuple(FROZEN_TRANSACTION_ID, INVALID_TRANSACTION_ID);
        let snapshot = Snapshot::new(100, 40, 150);

        let result = heap_tuple_satisfies_mvcc(&header, &snapshot, &oracle);
        assert_eq!(result, VisibilityResult::Visible);
    }

    #[test]
    fn delete_not_visible_to_old_snapshot() {
        let mut oracle = SimpleTransactionOracle::new();
        oracle.commit(50);
        oracle.commit(100); // Delete committed

        let header = make_tuple(50, 100);

        // Snapshot that can't see transaction 100
        let snapshot = Snapshot::new(80, 40, 90);

        let result = heap_tuple_satisfies_mvcc(&header, &snapshot, &oracle);
        assert_eq!(result, VisibilityResult::Visible);
    }
}

File: internal/storage/visibility.go
package storage

// MVCC tuple visibility (src/backend/access/heap/heapam_visibility.c).
//
// PostgreSQL determines whether a heap tuple is visible to a transaction by
// inspecting the tuple's xmin/xmax and comparing them against a Snapshot.
//
// Rules (simplified to the MVCC snapshot case):
//  1. xmin must be committed and visible to the snapshot.
//  2. xmax must be invalid, aborted, or not visible to the snapshot.
//
// Hint bits (HEAP_XMIN_COMMITTED, HEAP_XMAX_INVALID, etc.) cache the result
// of transaction status lookups so they aren't repeated on every tuple scan.

// TransactionStatus is the commit state of a transaction.
type TransactionStatus int

const (
	TxInProgress TransactionStatus = iota
	TxCommitted
	TxAborted
)

// TransactionOracle provides transaction status for visibility checks.
// In a real implementation this consults pg_xact (clog).
type TransactionOracle interface {
	Status(xid TransactionId) TransactionStatus
}

// Snapshot represents a consistent point-in-time view of the database.
//
// Matches PostgreSQL's SnapshotData (snapshot.h):
//   - Xmin: all XIDs < Xmin are visible (if committed)
//   - Xmax: all XIDs >= Xmax are invisible
//   - Xip:  in-progress XIDs between Xmin and Xmax (not visible)
//   - Xid:  the current transaction's own XID
type Snapshot struct {
	Xid    TransactionId   // current transaction ID
	Xmin   TransactionId   // oldest active XID at snapshot time
	Xmax   TransactionId   // first as-yet-unassigned XID
	Xip    []TransactionId // in-progress XIDs [Xmin, Xmax)
	CurCid CommandId       // current command ID within transaction
}

// NewSnapshot constructs a snapshot. xip may be nil.
func NewSnapshot(xid, xmin, xmax TransactionId) *Snapshot {
	return &Snapshot{Xid: xid, Xmin: xmin, Xmax: xmax}
}

// AddInProgress marks xid as in-progress in this snapshot.
func (s *Snapshot) AddInProgress(xid TransactionId) {
	for _, x := range s.Xip {
		if x == xid {
			return
		}
	}
	s.Xip = append(s.Xip, xid)
}

// XidVisible reports whether xid is visible in this snapshot.
//
// Logic mirrors XidInMVCCSnapshot() in PostgreSQL:
//   - FrozenTransactionId is always visible.
//   - InvalidTransactionId is never visible.
//   - The current transaction's own XID is visible.
//   - XIDs >= Xmax are invisible (committed after snapshot).
//   - XIDs < Xmin are visible (committed before snapshot).
//   - XIDs in [Xmin, Xmax) are visible unless they appear in Xip.
func (s *Snapshot) XidVisible(xid TransactionId) bool {
	if xid == InvalidTransactionId {
		return false
	}
	if xid == FrozenTransactionId {
		return true
	}
	if xid == s.Xid {
		return true
	}
	if xid >= s.Xmax {
		return false
	}
	if xid < s.Xmin {
		return true
	}
	// In [Xmin, Xmax) — visible only if not in Xip.
	for _, x := range s.Xip {
		if x == xid {
			return false
		}
	}
	return true
}

// VisibilityResult is the outcome of a tuple visibility check.
type VisibilityResult int

const (
	VisVisible       VisibilityResult = iota // tuple visible to snapshot
	VisInvisible                             // inserted by uncommitted/aborted xact
	VisDeleted                               // deleted and deletion is visible
	VisBeingDeleted                          // current xact is deleting it
	VisInsertedBySelf                        // inserted by current xact, visible
	VisDeletedBySelf                         // deleted by current xact
)

// IsVisible reports whether the result means the tuple should be returned.
func (r VisibilityResult) IsVisible() bool {
	return r == VisVisible || r == VisInsertedBySelf
}

// HeapTupleSatisfiesMVCC implements PostgreSQL's HeapTupleSatisfiesMVCC.
//
// It checks whether the tuple described by hdr is visible to snap using
// oracle for transaction status lookups.
func HeapTupleSatisfiesMVCC(hdr *HeapTupleHeader, snap *Snapshot, oracle TransactionOracle) VisibilityResult {
	infomask := hdr.Infomask()

	// ── xmin check ──────────────────────────────────────────────────────────
	if !xminVisible(hdr, snap, oracle, infomask) {
		return VisInvisible
	}

	// ── self-insert check ────────────────────────────────────────────────────
	if hdr.TXmin == snap.Xid {
		if hdr.TCid >= snap.CurCid {
			// Inserted by a later command within the same transaction.
			return VisInvisible
		}
		if hdr.TXmax == InvalidTransactionId || infomask&HeapXmaxInvalid != 0 {
			return VisInsertedBySelf
		}
		if hdr.TXmax == snap.Xid {
			return VisDeletedBySelf
		}
	}

	// ── xmax check ──────────────────────────────────────────────────────────
	return xmaxVisibility(hdr, snap, oracle, infomask)
}

// xminVisible reports whether xmin makes the tuple eligible for visibility.
func xminVisible(hdr *HeapTupleHeader, snap *Snapshot, oracle TransactionOracle, infomask InfomaskFlags) bool {
	xmin := hdr.TXmin
	if xmin == InvalidTransactionId {
		return false
	}
	if xmin == FrozenTransactionId {
		return true
	}

	// Consult hint bits first to avoid a clog lookup.
	if infomask&HeapXminCommitted != 0 {
		return snap.XidVisible(xmin)
	}
	if infomask&HeapXminInvalid != 0 {
		return false
	}

	// No hint bits; ask the oracle.
	switch oracle.Status(xmin) {
	case TxCommitted:
		return snap.XidVisible(xmin)
	case TxAborted:
		return false
	default: // TxInProgress
		return xmin == snap.Xid
	}
}

// xmaxVisibility determines the final visibility result once xmin is known visible.
func xmaxVisibility(hdr *HeapTupleHeader, snap *Snapshot, oracle TransactionOracle, infomask InfomaskFlags) VisibilityResult {
	xmax := hdr.TXmax
	if xmax == InvalidTransactionId || infomask&HeapXmaxInvalid != 0 {
		return VisVisible
	}

	// Hint bit: xmax already committed.
	if infomask&HeapXmaxCommitted != 0 {
		if snap.XidVisible(xmax) {
			return VisDeleted
		}
		return VisVisible
	}

	// Current transaction is the deleter.
	if xmax == snap.Xid {
		return VisBeingDeleted
	}

	switch oracle.Status(xmax) {
	case TxCommitted:
		if snap.XidVisible(xmax) {
			return VisDeleted
		}
		return VisVisible
	case TxAborted:
		return VisVisible
	default: // TxInProgress (another transaction is deleting)
		return VisVisible
	}
}

// ── Additional snapshot types ─────────────────────────────────────────────────

// HeapTupleSatisfiesDirty reports whether a tuple is the latest version of a
// row, regardless of commit status.  Used internally when examining the live
// tuple during an update (PostgreSQL's HeapTupleSatisfiesDirty).
//
// Rules:
//   - If xmin is aborted, the tuple never existed: invisible.
//   - If xmax is absent or aborted, this is the latest version: visible.
//   - Otherwise the tuple has been or is being deleted: invisible.
func HeapTupleSatisfiesDirty(hdr *HeapTupleHeader, oracle TransactionOracle) VisibilityResult {
	xmin := hdr.TXmin
	if xmin == InvalidTransactionId {
		return VisInvisible
	}
	if xmin != FrozenTransactionId && oracle.Status(xmin) == TxAborted {
		return VisInvisible
	}

	xmax := hdr.TXmax
	infomask := hdr.Infomask()
	if xmax == InvalidTransactionId || infomask&HeapXmaxInvalid != 0 {
		return VisVisible
	}
	if oracle.Status(xmax) == TxAborted {
		return VisVisible
	}
	return VisDeleted
}

// HeapTupleSatisfiesNow reports whether a tuple is visible to any reader that
// uses a snapshot taken right now (i.e., it is the most recent committed
// version).  Mirrors PostgreSQL's SnapshotNow semantics: xmin committed, xmax
// either absent or not yet committed.
func HeapTupleSatisfiesNow(hdr *HeapTupleHeader, oracle TransactionOracle) VisibilityResult {
	xmin := hdr.TXmin
	if xmin == InvalidTransactionId {
		return VisInvisible
	}

	infomask := hdr.Infomask()

	// xmin must be committed.
	xminCommitted := false
	if xmin == FrozenTransactionId {
		xminCommitted = true
	} else if infomask&HeapXminCommitted != 0 {
		xminCommitted = true
	} else if infomask&HeapXminInvalid != 0 {
		return VisInvisible
	} else {
		switch oracle.Status(xmin) {
		case TxCommitted:
			xminCommitted = true
		case TxAborted:
			return VisInvisible
		default:
			return VisInvisible // still in progress
		}
	}
	if !xminCommitted {
		return VisInvisible
	}

	// xmax must be absent or not yet committed.
	xmax := hdr.TXmax
	if xmax == InvalidTransactionId || infomask&HeapXmaxInvalid != 0 {
		return VisVisible
	}
	if infomask&HeapXmaxCommitted != 0 {
		return VisDeleted
	}
	switch oracle.Status(xmax) {
	case TxCommitted:
		return VisDeleted
	case TxAborted:
		return VisVisible
	default: // xmax in progress: tuple is being deleted but not yet gone
		return VisVisible
	}
}

// HeapTupleSatisfiesAny always returns VisVisible.  Used by VACUUM and index
// builds that need to see every tuple regardless of transaction state.
// Mirrors PostgreSQL's SnapshotAny.
func HeapTupleSatisfiesAny(_ *HeapTupleHeader) VisibilityResult {
	return VisVisible
}

// SetHintBits updates HEAP_XMIN_COMMITTED / HEAP_XMIN_INVALID and
// HEAP_XMAX_COMMITTED / HEAP_XMAX_INVALID hint bits based on oracle status.
// This should be called after a successful visibility check to avoid repeated
// clog lookups (write-once, then cheap to re-read).
func SetHintBits(hdr *HeapTupleHeader, oracle TransactionOracle) {
	infomask := hdr.Infomask()

	if infomask&HeapXminCommitted == 0 && infomask&HeapXminInvalid == 0 {
		switch oracle.Status(hdr.TXmin) {
		case TxCommitted:
			hdr.TInfomask = uint16(infomask | HeapXminCommitted)
			infomask = hdr.Infomask() // re-read
		case TxAborted:
			hdr.TInfomask = uint16(infomask | HeapXminInvalid)
			infomask = hdr.Infomask()
		}
	}

	if hdr.TXmax != InvalidTransactionId &&
		infomask&HeapXmaxCommitted == 0 && infomask&HeapXmaxInvalid == 0 {
		switch oracle.Status(hdr.TXmax) {
		case TxCommitted:
			hdr.TInfomask = uint16(infomask | HeapXmaxCommitted)
		case TxAborted:
			hdr.TInfomask = uint16(infomask | HeapXmaxInvalid)
		}
	}
}

// ── SimpleTransactionOracle ──────────────────────────────────────────────────

// SimpleTransactionOracle is an in-memory oracle backed by a map.
// Useful for unit tests and single-process experiments.
type SimpleTransactionOracle struct {
	statuses map[TransactionId]TransactionStatus
}

// NewSimpleOracle creates an empty oracle (unknown XIDs default to TxInProgress).
func NewSimpleOracle() *SimpleTransactionOracle {
	return &SimpleTransactionOracle{statuses: make(map[TransactionId]TransactionStatus)}
}

// SetStatus records the status for xid.
func (o *SimpleTransactionOracle) SetStatus(xid TransactionId, s TransactionStatus) {
	o.statuses[xid] = s
}

// Commit records xid as committed.
func (o *SimpleTransactionOracle) Commit(xid TransactionId) { o.SetStatus(xid, TxCommitted) }

// Abort records xid as aborted.
func (o *SimpleTransactionOracle) Abort(xid TransactionId) { o.SetStatus(xid, TxAborted) }

// Status implements TransactionOracle.
func (o *SimpleTransactionOracle) Status(xid TransactionId) TransactionStatus {
	if xid == InvalidTransactionId {
		return TxAborted
	}
	if xid == FrozenTransactionId {
		return TxCommitted
	}
	if s, ok := o.statuses[xid]; ok {
		return s
	}
	return TxInProgress
}
